beluga/include/beluga/actions/overlay.hpp

```diff
@@ -0,0 +1,130 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_ACTIONS_OVERLAY_HPP
+#define BELUGA_ACTIONS_OVERLAY_HPP
+
+#include <algorithm>
+#include <execution>
+
+#include <range/v3/action/action.hpp>
+#include <range/v3/view/common.hpp>
+#include <range/v3/view/transform.hpp>
+
+/**
+ * \file
+ * \brief Implementation of the overlay range adaptor object
+ */
+namespace beluga::actions {
+
+namespace detail {
+
+/// Implementation detail for an overlay range adaptor object.
+struct overlay_base_fn {
+  /// Overload that implements an overlay of a value in a range.
+  /**
+   * \tparam ExecutionPolicy An [execution policy](https://en.cppreference.com/w/cpp/algorithm/execution_policy_tag_t).
+   * \tparam Range An [input range](https://en.cppreference.com/w/cpp/ranges/input_range).
+   * \tparam MaskRange An [input range](https://en.cppreference.com/w/cpp/ranges/input_range).
+   * \param policy The execution policy to use.
+   * \param range An existing range to apply this action to.
+   * \param mask The mask where the values will be overlaid.
+   * \param mask_value The value to be overlaid.
+   */
+  template <
+      class ExecutionPolicy,
+      class Range,
+      class MaskRange,
+      class Mask = ranges::range_value_t<Range>,
+      std::enable_if_t<std::is_execution_policy_v<std::decay_t<ExecutionPolicy>>, int> = 0,
+      std::enable_if_t<ranges::range<Range>, int> = 0,
+      std::enable_if_t<ranges::range<MaskRange>, int> = 0>
+  constexpr auto operator()(ExecutionPolicy&& policy, Range& range, const MaskRange& mask, const Mask& mask_value) const
+      -> Range& {
+    auto map = range | ranges::views::common;
+
+    std::transform(
+        policy,            //
+        std::begin(map),   //
+        std::end(map),     //
+        std::begin(mask),  //
+        std::begin(map),   //
+        [mask_value](const auto& base_value, bool flag) { return flag ? mask_value : base_value; });
+
+    return range;
+  }
+
+  /// Overload that re-orders arguments from an action closure.
+  template <
+      class ExecutionPolicy,
+      class Range,
+      class MaskRange,
+      class Mask = ranges::range_value_t<Range>,
```

Comment:
```suggestion
      class Mask,
```

Reply:
If I remove this, how could I make `Mask` to be derived from `Range`? Discussed [here](https://github.com/Ekumen-OS/beluga/pull/430#discussion_r1770736903).

Reply:
So, if you don't remove this, the compiler won't deduce the type and `std::forward` won't work as expected.
If you want the conversion to only happen once, you can do this just before calling `std::transform`:

```cpp
const auto converted_mask_value = static_cast<ranges::range_value_t<Range>>(mask_value);
```

Reply:
Interesting. I haven't realized about all the details behind forwarding. Thanks @nahueespinosa, TIL!